apis/python/src/tiledbsoma/options/_soma_tiledb_context.py

```diff
@@ -128,6 +131,18 @@ def __init__(
         self._tiledb_ctx = tiledb_ctx
         """The TileDB context to use, either provided or lazily constructed."""
         self._timestamp_ms = _maybe_timestamp_ms(timestamp)
+        """Concurrency factor to be used by the C++ re-indexer"""
+        self.soma_indexer_concurrency = soma_indexer_concurrency
+        if soma_indexer_concurrency is not None:
+            clib.IntIndexer.start_thread_pool(soma_indexer_concurrency)
+        elif tiledb_ctx is not None:
+            tdb_concurrency = int(
+                tiledb_ctx.config().get("sm.compute_concurrency_level", 10)
+            )
+            thread_count = max(1, tdb_concurrency // 2)
+            clib.IntIndexer.start_thread_pool(thread_count)
+        else:
+            clib.IntIndexer.start_thread_pool(5)
```

Comment:
This and `__del__` lead to several race conditions—if I have two threads, and an indexer running in one, then starting a second one will lead to the following:

1. Start indexer A. It starts the global thread pool.
2. Indexer A runs.
3. Start indexer B. It starts a new global thread pool, thus deleting the old one.
4. Indexer B runs. (What happens to the threads of indexer A that are still running under the old pool?)
5. Indexer A completes and is garbage collected (assuming some other problem does not arise). It deletes the thread pool entirely in `__del__`.
6. Indexer B cannot continue to run. (Does it crash? Just freeze?)

Reply:
@thetorpedodog thanks for the comments. Thinking about them, some clarifications:

1. The thread pool lifetime is tied to creation and deletion of soma context and not the indexers
2. The thread pool supports parallel indexers with no race issues assuming they are both attached to the same context (it uses indexers as keys). I can add unittests for that.
3. I believe the only situation that can lead to the race condition you mentioned is when we have two parallel indexers each with a different context. In that case the scenario you mentioned can take place

Reply:
Already discussed in DMs but x-posting here: Instead of a global map, we should add `tiledb_threadpool` as a property of `TileDBSOMAContext`. It could also be lazy loaded just like how we do with `TileDBSOMAContext.tiledb_ctx`. Unless I'm misunderstanding something below, there isn't already a preexisting threadpool that can be directly accessed via `tiledb::Context` and has to be instantiated within `TileDBSOMAContext` in Python or the upcoming `SOMAContext` in C++.

Reply:
We have the start of a thread pool in somacore (`threadpool` member on the `ContextBase` protocol), but it seems to have not made it fully to TileDB-SOMA yet? https://github.com/single-cell-data/SOMA/blob/76b2d2bbca7c8baf543b5cc9cf71fc5c9138a39a/python-spec/src/somacore/types.py#L87

Reply:
OIC - that makes more sense now. Yeah `SOMATIleDBContext` doesn't inherit from `soma.ContextBase` yet, but that'll be easy to add and then we can start utilizing the context's threadpool.

Reply:
OK, for the time being I'm gonna implement the design with the thread on top of the C++ Soma context.